### PR TITLE
create: keep linear_attn in_proj_qkv and in_proj_z in BF16 for NVFP4/MXFP4/MXFP8

### LINKED ISSUE #15866: qwen3.6:35b-a3b-coding-nvfp4 model file has corrupted K-projection weights (layers 0-1 entirely zero), making linear attention output zero
# `qwen3.6:35b-a3b-coding-nvfp4` model file has corrupted K-projection weights for early linear-attention layers

## What is the issue?

The NVFP4 packaging of `qwen3.6:35b-a3b-coding-nvfp4` shipped from `registry.ollama.ai/library/qwen3.6` has the K-projection portion of `linear_attn.in_proj_qkv.weight` **entirely zeroed out** for layer 0 and mostly zeroed for layer 1 — making coherent generation impossible regardless of any runner-side fix.

This is a model-packaging issue, not a runtime bug. The `mlx-community/Qwen3.6-35B-A3B-4bit` model (same architecture, same source weights, MLX 4-bit affine quant instead of NVFP4) has the expected natural sparsity (3-5%) in the same regions.

## Reproducer

The `linear_attn.in_proj_qkv.weight` tensor is laid out per HF transformers as `[Q rows 0..2047 | K rows 2048..4095 | V rows 4096..8191]`. A direct read of the safetensors blob shows:

```
layer  0  K-zero-rows = 2048 / 2048   (100% — entire K projection is zero!)
layer  1  K-zero-rows = 1797 / 2048   ( 88%)
layer  2  K-zero-rows =   57 / 2048   (  3%)
layer  3  ... (full attention layer, no in_proj_qkv)
layer  4+ K-zero-rows ≤ 2 / 2048      (~natural sparsity)
```

For comparison, on `mlx-community/Qwen3.6-35B-A3B-4bit` (4-bit affine, same architecture):
```
layer 0  K-zero-rows =  56 / 2048
layer 1  K-zero-rows = 102 / 2048
layer 2  K-zero-rows =  56 / 2048
```

The mlx-community model generates coherent text via `mlx-lm generate` (top next token for "What is 2+2?" → "Here", correctly leading into a thinking trace). The Ollama NVFP4 model cannot generate coherent text under any code fix because layers 0 and 1 of the linear-attention path produce K=0 → linear attention output = 0 → the residual stream loses the entire linear-attention contribution from those layers.

## Verification commands

```python
# Reproduce against the cached blob (replace the SHA with whichever blob holds the layer-0 in_proj_qkv tensor):
import struct, json, numpy as np
blob = "/Users/<you>/.ollama/models/blobs/sha256-..."   # blob carrying layers.0.linear_attn.in_proj_qkv
with open(blob, 'rb') as f:
    hsz = struct.unpack('<Q', f.read(8))[0]
    hdr = json.loads(f.read(hsz))
    body = 8 + hsz
    info = hdr['language_model.model.layers.0.linear_attn.in_proj_qkv.weight']
    s, e = info['data_offsets']
    f.seek(body + s)
    arr = np.frombuffer(f.read(e-s), dtype=np.uint32).reshape(info['shape'])
    print('K rows zero:', sum(1 for r in arr[2048:4096] if not r.any()), '/ 2048')
# → 2048 / 2048 on the broken model
```

## Likely cause

The conversion from BF16 (or the source FP8/BF16 RedHatAI NVFP4) to Ollama's NVFP4 packaging picked a global scale per tensor that was too large to represent the small magnitudes in early-layer K projections — they all rounded to the FP4 zero codepoint.

For verification: the *scales* tensor for `in_proj_qkv` is non-zero across all rows (max-row-abs ≈ 1144 on K rows for layer 0), but the underlying FP4 nibbles are all zero, so dequantization yields zero regardless of the (non-zero) scale. That fingerprint matches "small-magnitude rows quantized below FP4 representable range" rather than "block of weights deleted in I/O".

## Suggested fixes

1. Re-quantize with per-tensor (or per-row block) scales chosen to preserve the K projection's dynamic range — or fall back to a higher-precision quant for early-layer linear-attention weights.
2. Check the same artefact for the `mxfp8` and other variants in the library (`qwen3.6:35b-a3b-coding-mxfp8`, `qwen3.6:35b-a3b-nvfp4`, `qwen3.6:27b-coding-nvfp4`).
3. Re-run the conversion using the per-tensor quant overrides path (e.g. matching the pattern in #15760, where `mlp.gate` and `mlp.shared_expert_gate` get an 8-bit override) — applying an 8-bit override (or skipping quant entirely) to `linear_attn.in_proj_qkv` would avoid the zero-collapse.

## Related

* #15865 — independent recurrent-state precision bug in `gated_delta_step` (the kernel cast state to `InT` instead of `StT`). With **only** that fix and **without** this corruption, the model reaches the sampler but the very first decode step gets wrong logits because layer-0 linear attention contributes zero. Together, both issues explain the "broken Qwen3.6 NVFP4 generation on Ollama".
* #15822, #15834, #15700 — user reports of broken generation on this model family on macOS; this issue likely explains the coding-nvfp4 sub-cases.
* mlx-community / Qwen3.6-35B-A3B-4bit on HF works correctly via `mlx-lm`, demonstrating the architecture is otherwise sound.

## Environment

* Ollama version: v0.22.0 (as of 2026-04-28)
* OS: macOS 26.4.1, Apple M4 Max
* Source model: `registry.ollama.ai/library/qwen3.6:35b-a3b-coding-nvfp4` (digest `sha256:cd2692a833e6...`)

### PR DESCRIPTION
**Pull request by ArkaD171717**
## Summary

Adds `in_proj_qkv.weight` and `in_proj_z.weight` to the BF16 exemption
list in `qwen35ShouldKeepBF16ForDirectNonAffine`. The upstream source
(RedHatAI/Qwen3.6-35B-A3B-NVFP4) keeps all `linear_attn` projections in
BF16 -- `in_proj_a` and `in_proj_b` were already exempted but `in_proj_qkv`
and `in_proj_z` were missing, so they got NVFP4-quantized.

The K-projection rows in `in_proj_qkv` have small magnitudes (0.01-0.04)
in early layers that fall below the smallest FP4 codepoint at group_size=16,
zeroing out layers 0-1 entirely.

### Weight magnitude analysis (source BF16, Qwen/Qwen3.6-35B-A3B)

| Tensor | Layer | Median abs | % < 0.01 | % < 0.05 |
|--------|-------|-----------|----------|----------|
| in_proj_qkv | 0 | 0.0093 | 52.7% | 99.5% |
| in_proj_z | 0 | 0.0109 | 46.3% | 99.4% |
| in_proj_qkv | 1 | 0.0100 | 49.8% | 99.4% |
| in_proj_z | 1 | 0.0101 | 49.4% | 99.5% |

Both tensors have ~50% of weights below the smallest FP4 codepoint at group_size=16. in_proj_z has the same zero-collapse risk as in_proj_qkv.

Test updated, passes locally for nvfp4/mxfp8/mxfp4.

Fixes #15866

### CODE REVIEW

**Comment by BruceMacD** [comment]
Issue #15866 only documents corruption on `in_proj_qkv`. Did you actually observe zero-collapse on `in_proj_z`, or is this added purely for parity with upstream RedHatAI keeping all `linear_attn` projections in BF16? Worth stating in the description either way.

**Comment by ArkaD171717** [comment]
Hello, thank you for the reply!
I actually looked into this earlier and checked the source BF16 weights via range requests on the safetensors file
in_proj_z has the same small-magnitude profile as in_proj_qkv in early layers:
- for layer 0 in_proj_qkv: median abs 0.0093, 52.7% of weights < 0.01
- layer 0 in_proj_z: median abs 0.0109, 46.3% of weights < 0.01
- layer 1 in_proj_qkv: median abs 0.0100, 49.8% of weights < 0.01
- layer 1 in_proj_z: median abs 0.0101, 49.4% of weights < 0.01

Both have ~99.5% of values below 0.05 w/ ~50% of weights below the smallest FP4 codepoint at group_size=16 which is same zero-collapse risk as in_proj_qkv

Just forgot to update the PR description. It'll be done soon to include this data

**Comment by ArkaD171717** [comment]
Hey @pdevine, any chance you could take a quick look at this? Been open for 2 weeks. Thanks

**Comment by ArkaD171717** [comment]
@xmddmx confirmed in #15880 that q4_K_M completes fine on 0.30.0-rc17 but nvfp4 still hangs indefinitely

**Comment by pdevine** [comment]
if you look at the `q4_K_M` model, it converts both the `in_proj_qkv` and `in_proj_z` into 4 bit integer format:

```
    blk.0.attn_gate.weight                     Q4_K    [2048 4096]
    blk.0.attn_qkv.weight                      Q4_K    [2048 8192]
```

Converting these to BF16 would cause the model to be more bloated and slow down, so I don't believe this change is the answer. I'm going to keep looking though; I'm not convinced MLX is correctly setting the global scale for the nvfp4 tensors which could result in lower precision.

**Comment by pdevine** [comment]
Definitely appreciate the PR, but I'm going to go ahead and close this as those tensors should be quantized.
